flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeStore.java

```diff
@@ -136,21 +133,23 @@ private void drainAndSave() {
             tasks = new ArrayList<>(scheduled);
             scheduled.clear();
             scheduledSizeInBytes = 0;
-            scheduledFuture = null;
         }
         try {
-            if (error != null) {
-                tasks.forEach(task -> task.fail(error));
+            if (error.get() != null) {
+                tasks.forEach(task -> task.fail(error.get()));
                 return;
             }
-            retryingExecutor.execute(retryPolicy, () -> delegate.save(tasks));
+            delegate.save(tasks);
+
+            synchronized (scheduled) {
+                scheduleUploadIfNeeded();
+            }
         } catch (Throwable t) {
             tasks.forEach(task -> task.fail(t));
             if (findThrowable(t, IOException.class).isPresent()) {
                 LOG.warn("Caught IO exception while uploading", t);
             } else {
-                error = t;
-                throw t;
+                error.compareAndSet(null, t);
```

Comment:
Could you explain the removal of `throw t;` ?
Without it:
1. error can be not logged at all - if no more upload scheduled
2. or if upload errors are not logged later
3. it can be logged with a delay (complicated debug)

Reply:
This is a top level method from the uploader/executor thread pool and the only meaningful way of returning the error to the task thread is via `error`? If we throw here, it would have no effect, or am I missing something?

Reply:
You're right (in my version, the method was also called directly).